ddprof-lib/src/main/cpp/threadFilter.cpp

```diff
@@ -1,175 +1,271 @@
-  _enabled = false;
-  _size = 0;
+#include <cstdlib>
+#include <thread>
+#include <algorithm>
+#include <cstring>
+
+ThreadFilter::ShardHead ThreadFilter::_free_heads[ThreadFilter::kShardCount] {};
+
+ThreadFilter::ThreadFilter() : _enabled(false) {
+    // Initialize chunk pointers to null (lazy allocation)
+    for (int i = 0; i < kMaxChunks; ++i) {
+        _chunks[i].store(nullptr, std::memory_order_relaxed);
+    }
+    _free_list = std::make_unique<FreeListNode[]>(kFreeListSize);
+
+    // Initialize the first chunk
+    initializeChunk(0);
+    clear();
 }
 
 ThreadFilter::~ThreadFilter() {
-  for (int i = 0; i < _max_bitmaps; i++) {
-    if (_bitmap[i] != NULL) {
-      OS::safeFree(_bitmap[i], BITMAP_SIZE);
+    // Clean up allocated chunks
+    for (int i = 0; i < kMaxChunks; ++i) {
+        ChunkStorage* chunk = _chunks[i].load(std::memory_order_relaxed);
+        delete chunk;
     }
-  }
-  if (_bitmap) {
-    OS::safeFree(_bitmap, _max_bitmaps * sizeof(u64 *));
-  }
 }
 
-void ThreadFilter::init(const char *filter) {
-  if (filter == NULL) {
-    _enabled = false;
-    return;
-  }
-
-  char *end;
-  do {
-    int id = strtol(filter, &end, 0);
-    if (id <= 0) {
-      break;
-    }
+void ThreadFilter::initializeChunk(int chunk_idx) {
+    if (chunk_idx >= kMaxChunks) return;
+
+    // Check if chunk already exists
+    ChunkStorage* existing = _chunks[chunk_idx].load(std::memory_order_acquire);
+    if (existing != nullptr) return;
 
-    if (*end == '-') {
-      int to = strtol(end + 1, &end, 0);
-      while (id <= to) {
-        add(id++);
-      }
+    // Allocate new chunk
+    ChunkStorage* new_chunk = new ChunkStorage();
+
+    // Try to install it atomically
+    ChunkStorage* expected = nullptr;
+    if (_chunks[chunk_idx].compare_exchange_strong(expected, new_chunk, std::memory_order_acq_rel)) {
+        // Successfully installed - initialize all slots
+        for (auto& slot : new_chunk->slots) {
+            slot.value.store(-1, std::memory_order_relaxed);
+        }
+        new_chunk->initialized.store(true, std::memory_order_release);
     } else {
-      add(id);
+        // Another thread beat us to it - clean up our allocation
+        delete new_chunk;
+    }
+}
+
+ThreadFilter::SlotID ThreadFilter::registerThread() {
+    // First, try to get a slot from the free list (lock-free stack)
+    SlotID reused_slot = popFromFreeList();
+    if (reused_slot >= 0) {
+        return reused_slot;
+    }
+
+    // Allocate a new slot
+    SlotID index = _next_index.fetch_add(1, std::memory_order_relaxed);
+    if (index >= kMaxThreads) {
+        // Revert the increment and return failure
+        _next_index.fetch_sub(1, std::memory_order_relaxed);
+        return -1;
     }
 
-    filter = end + 1;
-  } while (*end);
+    const int chunk_idx = index >> kChunkShift;
 
-  _enabled = true;
+    // Ensure the chunk is initialized (lock-free)
+    if (chunk_idx >= _num_chunks.load(std::memory_order_acquire)) {
```

Comment:
I don't quite understand, are `index` and `chunk_idx` 1-to-1 matched?

Reply:
no, chunks are really groups of indices that will grow if we have a large amount of threads. The idea is to have roughly chunks of 256 threads and to grow if needed